src/App.tsx

```diff
@@ -1,24 +1,283 @@
 
-const USER_ID = 0;
+const USER_ID = 6725;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [isHeaderDisabled, setIsHeaderDisabled] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo>();
+  const [processings, setProcessings] = useState<number[]>([]);
+  const [currentFilter, setCurrentFilter] = useState<Filter>(Filter.all);
+
+  const removeMessage = () => {
+    setErrorMessage('');
+  };
+
+  const createErrorMessage = (title: ErrorsMessages) => {
+    setErrorMessage(title);
+  };
+
+  const showErrorMessage = (message: ErrorsMessages) => {
+    createErrorMessage(message);
+    setTimeout(() => removeMessage(), 3000);
+  };
+
+  const loadTodos = async () => {
+    setErrorMessage('');
+
+    try {
+      const dataFromServer = await getTodos();
+
+      setTodos(dataFromServer);
+    } catch (error) {
+      showErrorMessage(ErrorsMessages.Load);
+    }
+  };
+
+  const setFilter = (filter: Filter) => {
+    setCurrentFilter(filter);
+  };
+
+  const filterTodos = (filteringProperty: Filter) => {
+    return (
+      filteringProperty === Filter.all
+        ? todos
+        : todos.filter(todo => {
+          switch (filteringProperty) {
+            case Filter.active:
+              return !todo.completed;
+            case Filter.completed:
+              return todo.completed;
+            default:
+              return null;
+          }
+        })
+    );
+  };
+
+  const filteredTodos: Todo[] = useMemo(
+    () => filterTodos(currentFilter),
+    [todos, currentFilter],
+  );
+
+  const addTodo = async (newTodoTitle: string) => {
+    setErrorMessage('');
+    setIsHeaderDisabled(true);
+
+    try {
+      setTempTodo({
+        id: 0,
+        userId: USER_ID,
+        title: newTodoTitle,
+        completed: false,
+      });
+      const newTodoPost = await createTodo({
+        title: newTodoTitle,
+        userId: USER_ID,
+        completed: false,
+      });
+
+      setTodos((items) => [...items, newTodoPost]);
+    } catch (error) {
+      showErrorMessage(ErrorsMessages.Post);
+    }
+
+    setTempTodo(undefined);
+    setIsHeaderDisabled(false);
+  };
+
+  const filterTodosById = (todoElements: Todo[], id: number) => {
+    return todoElements.filter((item) => item.id !== id);
+  };
+
+  const removeTodo = async (id: number) => {
+    setErrorMessage('');
+    setProcessings((prev) => [...prev, id]);
+
+    try {
+      const response = await deleteTodo(id);
+
+      if (response) {
+        setTodos(filterTodosById(todos, id));
+      }
+    } catch (error) {
+      showErrorMessage(ErrorsMessages.Delete);
+    }
+
+    setProcessings([]);
+  };
+
+  const removeCompletedTodos = async () => {
+    setErrorMessage('');
+
+    const completedTodos = (
+      todos.filter(el => el.completed)
+        .map(item => item.id)
+    );
+
+    setProcessings(completedTodos);
```

Comment:
but why & for What?

Reply:
if you are working with async/await structure it's the best practice to do it like that 

```
const func = async () => {
  try {
    setLoading(true);
    
    // other commenads
    
    const res = await request(body);
    
    setResult(res);
  } catch(err) {
    setError(err.message || 'Oops, smth went wrong. Try again pls')
  } finally {
    setLoading(false)
  }
}
```

Reply:
it's better for user to see the loading right away because there coud be cases when preparing the request body object could take some time as well , yes it's just milliseconds, but anyway)

Reply:
and there could be the case when error is caused by your code, not by the failed request, some syntax errors and so on - so it would not break the UX - user will se the error anyway and will be able to report it or just will be aware that his action was failed

Reply:
Ok, thanks